Daylight savings

huss opened the pull request:
# Description

eGauge meters are the first OED has acquired that honor daylight savings time. This leads to issues as time goes forward and backward during the shifts. This PR fixes it so times and reading values are correct. See the devDocs for the ideas and design of these changes. Additional, manual tests were added for the pipeline. 

This also added values associated with meters. Also, the start and end date/time on the meter is now a string instead of a moment object to preserve the offset. These are added to the CSV page and for curl-type commands.

Other changes made:
- Fixed old pipeline bug where reported non-cumulative readings with negative value as an error.
-  Fixed previous bug where import of MAMAC meters failed due to incorrect rename of ip to url.
- Other changes are noted in the commit messages.

Note this is a branch off off eGauge so it should not be merged until PR #764 is merged. Those changes should be reviewed as part of that PR.

## Type of change

- [ ] Note merging this changes the node modules
- [X] Note merging this changes the database configuration.
- [X] This change requires a documentation update

## Checklist

- [X] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [X] I have removed text in ( ) from the issue request

## Limitations

- This only works if the timestamp includes an offset. There is no plan to undo this due to potential issues.
- It detects but does not correct for readings that go back in time when crossing out of DST. This is viewed as okay and will stay unless users have issues.

huss committed:
pipeline understands DST +

This is a good first pass at making the pipeline understand daylight
savings time. It still needs careful review and testing. It also does
not work if the offset is not given in the date/timestamp. It also
is not yet integrated into the full pipeline including CSV uploads.
The major changes are:

- pipeline can deal with DST shifts. See devDoc for more details.
- pipeline changed in how parses date/time strings due to other changes.
The user must now state if want a more relaxed parsing.
- Fixed old pipeline bug where reported non-cumulative readings with
negative value as an error.
- The DB meters table now has the start/end_timestamp to be a string
because that preserves the offset. There is also a new column to
keep the previous end timestamp for when the pipeline has batches of
readings that are separated while processing a cross out of DST.
- When getting meters by id, name use * so always get all columns.
- Tests updated for new meter columns.
- The expected output from the pipeline tests was updated. First, you
get +00:00 not Z for start/end timestamp. Second, the output
when questionable date/time formats are given was changed.
- src/server/test/web/csvPipeline/regAscVariedFormats.csv was changed
to remove the case where the year was not first with hyphens
since the new use of moment does not support this.

Other changes:
- A typo in src/server/models/Meter.js for rul -> url.
- src/server/services/meterTimezone.js changed to use !==.
- src/server/services/csvPipeline/saveCsv.js did not have require
for moment
- src/server/test/db/configfileTests.js checked dates by toISOString()
and changed to moment.format() because had issue where looked at
fractions of a second and it differed.
- src/server/test/web/csvPipelineTest.js used the _i field of moment but
that is not recommended and caused issues. Used format instead.
- src/server/test/web/meters.js fixed for timezone format.
- src/server/test/db/meterTests.js fixed for new way timestamp works
and for default values.

huss committed:
DST, relaxed works with CSV upload + other

This adds the options to the CSV upload page. It also passes the new
parameters to the pipeline in all uses.
- Updated eGauge getting of meter timestamp since now string.
- Fixed previous bug where import of MAMAC meters failed due to
incorrect rename of ip to url.
- Removed old console.log that was commented out.

huss committed:
adds tests for daylight for pipeline +

The tests are manual as they currently are.

- Changed default value for new meter for previousEnd so not string.

huss committed:
deal with gap in readings

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into daylightSavings

- Fixed bug for end only where used wrong index in relaxedParsing.
- Updated cvsPipelineTest for timezone.

huss committed:
add note to automated test Readme

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/huss/781

huss commented:
I'm noting the build for the testing failed after the merge of development. It is unclear why this started. I did run the checks and tests on a local OED container and they all passed.

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/huss/781

- Fixed merge conflicts due to units
- Fixed tests for merge

huss commented:
I updated so in sync with development with units. I rechecked that works correctly with changes including moment package update. All seems fine.

truongdd03 commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -69,16 +73,57 @@ async function processData(rows, meterID, timeSort = TimeSortTypesJS.increasing,
+	let startTimestampTz = moment.parseZone(meter.startTimestamp, true);
+	let endTimestampTz = moment.parseZone(meter.endTimestamp, true);
+	startTimestamp = moment.parseZone(startTimestampTz.clone()).tz('UTC', true);
+	endTimestamp = moment.parseZone(endTimestampTz.clone()).tz('UTC', true);
```
Did you accidentally drop `let` before `endTimestamp`? The same with `startTimestamp` and `endTimestampMeterZone`.

huss replied:
Thanks. I have added the lets. I think I originally had them declared about during earlier work and then removed it so the let was lost. Good catch.
I debated what to do about the endTimestampMeterZone since it is used in three places. I decided to put a const in front of each use inside each if rather than a let at a higher scope.

truongdd03 commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -335,7 +506,80 @@ async function processData(rows, meterID, timeSort = TimeSortTypesJS.increasing,
+				// Remove that in DST split since done processing this reading and don't want for next one.
+				splitDst = false;
+			} else if (inDstStop) {
+				{
```
I think this bracket is redundant.

huss replied:
Thanks. Removed.

truongdd03 commented on src/server/migrations/0.8.0-1.0.0/sql/meters/alter_meter_time_holders.sql:
```diff
@@ -0,0 +1,13 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I think we should add this file to `src/server/migrations/0.8.0-1.0.0/index.js`. Also, I notice that 
a trailing comma in [line 16](https://github.com/OpenEnergyDashboard/OED/blob/32075ede2e1b95a6241ad23d916b7414be85087f/src/server/migrations/0.8.0-1.0.0/sql/unit/create_units_table.sql#L16) of `src/server/migrations/0.8.0-1.0.0/sql/unit/create_units_table.sql` leads to a syntax error. Could you please help me remove it?

huss replied:
I fixed the comma in units table creation.

I had previosly noticed we have a meters/ and meter/ migration. I moved this one to meter/ to be consistent and then added to the index.
For the record, I also dealt with the duplicated alter_meter_ipAddress.sql.

truongdd03 commented on src/server/services/pipeline-in-progress/processData.js:
```diff
@@ -126,38 +174,66 @@ async function processData(rows, meterID, timeSort = TimeSortTypesJS.increasing,
-			startTimestamp = moment.parseZone(rows[index][1], undefined, true).tz('UTC', true);
-			endTimestamp = moment.parseZone(rows[index][2], undefined, true).tz('UTC', true);
+			// Get the reading into moment in a timezone aware way. Note will assume UTC if no timezone in the string.
+			startTimestampTz = moment.parseZone(rows[index][1], undefined, true);
```
I find that the export format of OED is `Label,Readings,Start Timestamp,End Timestamp`. `rows[index][1]` is not `Start Timestamp`, hence it fails when I upload an exported file to OED. Is it an issue?

huss replied:
If I understand what you are concerned about the it involves the fact that the export and import formats do not match. You are correct about the difference. This is not a new issue. OED did the export for all meters together and lables each row with the meter name. The import is per meter without this column. This is planned to be fixed as part of the changes for exports to include units. The export will put each meter into its own file without the first column with the meter name. This is discussed in the resourceGeneralization document.

truongdd03 reviewed (commented):
Thank you @huss for working on this issue. I just have a couple of comments below.

huss committed:
address PR comments

huss committed:
fix meter pages for daylight savings

The meter pages needed updates:
- Added previousEnd that is new item.
- Make start/endTimestamp be a string rather than moment.

huss commented:
I'm updating the PR because I realized I had not updated the meter pages for the daylight savings changes.

huss commented:
This passes local testing so should now be fine for merging. I thank @truongdd03 for his review and that is being counted as the formal one for merging.
